cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/config/BrokerSetFileResolver.java

```diff
@@ -59,12 +59,25 @@ public Map<String, Set<Integer>> brokerIdsByBrokerSetId(ClusterModel clusterMode
     try {
       brokerIdsByBrokerSetId = loadBrokerSetData();
     } catch (IOException e) {
-      throw new IllegalArgumentException(e);
+      throw new BrokerSetResolutionException(e.getMessage());
     }
 
     return _brokerSetAssignmentPolicy.assignBrokerSetsForUnresolvedBrokers(clusterModel, brokerIdsByBrokerSetId);
   }
 
+  @Override
+  public Map<String, Set<Integer>> brokerIdsByBrokerSetId(Map<Integer, String> rackIdByBrokerId)
+      throws BrokerSetResolutionException {
+    Map<String, Set<Integer>> brokerIdsByBrokerSetId;
+    try {
+      brokerIdsByBrokerSetId = loadBrokerSetData();
+    } catch (IOException e) {
+      throw new BrokerSetResolutionException(e.getMessage());
+    }
+
+    return _brokerSetAssignmentPolicy.assignBrokerSetsForUnresolvedBrokers(rackIdByBrokerId, brokerIdsByBrokerSetId);
```

Comment:
Why are we not persisting this data back to the file? That would make sure that the unassigned brokers don't change colors on broker restart.

Reply:
@wyuka I think your proposal is reasonable, but there might be some other considerations when @mohitpali implemented this. @mohitpali do you think it would work better if we write back to the file?

In anyways, that part should belong to a new issue/PR. We want to focus this PR on the exact issue it linked.

Reply:
This was done to address a case where a broker is added to the cluster. Ideally the brokerSet to Broker mapping should be in-sync with the Cluster. But since the broker set mapping, even if dynamic, is not guaranteed to be updated atomically. In case of file based implementation, user should be updating the mapping json or implementing a resolution policy that is deterministic. I don't mind overwriting the file if user has saved the file at a non-ephemeral disk. But since this json is part of the code base, i would assume a new deployment would overwrite it.

Keeping the mapping read only makes more sense to me. For unmapped brokers, user could define a policy that is deterministic.